File: src/runtime/metadataResolver.js
function isPlainObject(value) {
    return !!value && typeof value === 'object' && !Array.isArray(value);
}

function cloneValue(value) {
    if (Array.isArray(value)) {
        return value.map((item) => cloneValue(item));
    }
    if (isPlainObject(value)) {
        const result = {};
        for (const [key, item] of Object.entries(value)) {
            result[key] = cloneValue(item);
        }
        return result;
    }
    return value;
}

function normalizeTargetSpec(spec) {
    if (!spec) return null;
    if (typeof spec === 'string') {
        const platform = spec.trim();
        if (!platform) return null;
        return {
            platforms: [platform],
            excludePlatforms: [],
            formFactors: [],
            capabilities: []
        };
    }
    if (Array.isArray(spec)) {
        const normalized = {
            platforms: spec.map((item) => typeof item === 'string' ? item.trim() : '').filter(Boolean),
            excludePlatforms: [],
            formFactors: [],
            capabilities: []
        };
        return hasTargetCriteria(normalized) ? normalized : null;
    }
    if (!isPlainObject(spec)) return null;
    const normalized = {
        platforms: trimmedStringList(spec.platforms),
        excludePlatforms: trimmedStringList(spec.excludePlatforms),
        formFactors: trimmedStringList(spec.formFactors),
        capabilities: trimmedStringList(spec.capabilities)
    };
    return hasTargetCriteria(normalized) ? normalized : null;
}

function trimmedStringList(value) {
    if (typeof value === 'string') {
        const trimmed = value.trim();
        return trimmed ? [trimmed] : [];
    }
    if (!Array.isArray(value)) return [];
    return value.map((item) => typeof item === 'string' ? item.trim() : '').filter(Boolean);
}

function hasTargetCriteria(normalized) {
    return normalized.platforms.length > 0 ||
        normalized.excludePlatforms.length > 0 ||
        normalized.formFactors.length > 0 ||
        normalized.capabilities.length > 0;
}

function matchesTarget(spec, targetContext = {}) {
    const normalized = normalizeTargetSpec(spec);
    if (!normalized) return true;

    const platform = String(targetContext.platform || '').trim();
    const formFactor = String(targetContext.formFactor || '').trim();
    const capabilities = new Set(
        Array.isArray(targetContext.capabilities)
            ? targetContext.capabilities.map((item) => String(item || '').trim()).filter(Boolean)
            : []
    );

    if (normalized.platforms.length > 0) {
        if (!platform || !normalized.platforms.includes(platform)) {
            return false;
        }
    }
    if (normalized.excludePlatforms.length > 0 && platform && normalized.excludePlatforms.includes(platform)) {
        return false;
    }
    if (normalized.formFactors.length > 0) {
        if (!formFactor || !normalized.formFactors.includes(formFactor)) {
            return false;
        }
    }
    if (normalized.capabilities.length > 0) {
        for (const capability of normalized.capabilities) {
            if (!capabilities.has(capability)) {
                return false;
            }
        }
    }
    return true;
}

function deepMerge(base, override) {
    if (Array.isArray(override)) {
        return override.map((item) => cloneValue(item));
    }
    if (!isPlainObject(override)) {
        return cloneValue(override);
    }
    const result = isPlainObject(base) ? cloneValue(base) : {};
    for (const [key, value] of Object.entries(override)) {
        const current = result[key];
        if (isPlainObject(current) && isPlainObject(value)) {
            result[key] = deepMerge(current, value);
            continue;
        }
        result[key] = deepMerge(undefined, value);
    }
    return result;
}

function applicableOverrides(targetOverrides, targetContext = {}) {
    if (!isPlainObject(targetOverrides)) return [];
    const keys = targetOverrideKeys(targetContext);
    const seen = new Set();
    const result = [];
    for (const key of keys) {
        if (!key || seen.has(key)) continue;
        seen.add(key);
        const candidate = targetOverrides[key];
        if (isPlainObject(candidate)) {
            result.push(candidate);
        }
    }
    return result;
}

function targetOverrideKeys(targetContext = {}) {
    const platform = String(targetContext.platform || '').trim();
    const formFactor = String(targetContext.formFactor || '').trim();
    const surface = String(targetContext.surface || '').trim();
    const mobilePlatforms = new Set(['android', 'ios']);
    const mobileFormFactors = new Set(['phone', 'tablet', 'foldable']);
    const isMobile = mobilePlatforms.has(platform) || mobileFormFactors.has(formFactor);
    const keys = [];

    if (surface) {
        keys.push(`surface:${surface}`);
        keys.push(surface);
    }
    if (isMobile) {
        keys.push('mobile');
    }
    if (formFactor) {
        keys.push(`formFactor:${formFactor}`);
        keys.push(formFactor);
    }
    if (platform) {
        keys.push(platform);
    }
    if (isMobile && formFactor) {
        keys.push(`mobile.${formFactor}`);
        keys.push(`mobile:${formFactor}`);
        keys.push(`mobile/${formFactor}`);
    }
    if (platform && formFactor) {
        keys.push(`${platform}.${formFactor}`);
        keys.push(`${platform}/${formFactor}`);
        keys.push(`${platform}:${formFactor}`);
        keys.push(platformFormFactorAlias(platform, formFactor));
    }
    return keys;
}

function platformFormFactorAlias(platform, formFactor) {
    return `${platform}${formFactor.charAt(0).toUpperCase()}${formFactor.slice(1)}`;
}

function resolveValue(node, targetContext = {}) {
    if (Array.isArray(node)) {
        const result = [];
        for (const item of node) {
            const resolved = resolveValue(item, targetContext);
            if (typeof resolved !== 'undefined') {
                result.push(resolved);
            }
        }
        return result;
    }
    if (!isPlainObject(node)) {
        return node;
    }

    if (!matchesTarget(node.target, targetContext)) {
        return undefined;
    }

    let working = cloneValue(node);
    const overrides = applicableOverrides(working.targetOverrides, targetContext);
    for (const override of overrides) {
        working = deepMerge(working, override);
    }

    if (normalizeTargetSpec(working.target)) {
        delete working.target;
    }
    if (
        isPlainObject(working.targetOverrides) &&
        Object.keys(working.targetOverrides).length > 0 &&
        Object.values(working.targetOverrides).every((value) => isPlainObject(value))
    ) {
        delete working.targetOverrides;
    }

    const result = {};
    for (const [key, value] of Object.entries(working)) {
        const resolved = resolveValue(value, targetContext);
        if (typeof resolved !== 'undefined') {
            result[key] = resolved;
        }
    }
    return result;
}

export function resolveMetadataForTarget(metadata, targetContext = {}) {
    if (!metadata || typeof metadata !== 'object') {
        return metadata;
    }
    return resolveValue(metadata, targetContext);
}

export function __private__() {
    return {
        normalizeTargetSpec,
        matchesTarget,
        deepMerge,
        applicableOverrides,
        targetOverrideKeys,
        resolveValue
    };
}

File: ios/Sources/ForgeIOSRuntime/MetadataResolver.swift
import Foundation

public enum MetadataResolver {
    public static func resolve(
        _ metadata: WindowMetadata,
        for targetContext: ForgeTargetContext
    ) -> WindowMetadata {
        guard
            let json = try? JSONEncoder().encode(metadata),
            let raw = try? JSONDecoder().decode(JSONValue.self, from: json),
            let resolved = resolveValue(raw, for: targetContext),
            let resolvedJSON = try? JSONEncoder().encode(resolved),
            let decoded = try? JSONDecoder().decode(WindowMetadata.self, from: resolvedJSON)
        else {
            return metadata
        }
        return decoded
    }

    private static func resolveValue(_ value: JSONValue, for targetContext: ForgeTargetContext) -> JSONValue? {
        switch value {
        case .array(let items):
            return .array(items.compactMap { resolveValue($0, for: targetContext) })
        case .object(let object):
            return resolveObject(object, for: targetContext)
        default:
            return value
        }
    }

    private static func resolveObject(
        _ object: [String: JSONValue],
        for targetContext: ForgeTargetContext
    ) -> JSONValue? {
        if !matchesTarget(object["target"], for: targetContext) {
            return nil
        }

        var working = object
        for override in applicableOverrides(object["targetOverrides"], for: targetContext) {
            working = deepMerge(base: working, override: override)
        }

        var result: [String: JSONValue] = [:]
        for (key, value) in working {
            if shouldStripTargetingKey(key, value) {
                continue
            }
            if let resolved = resolveValue(value, for: targetContext) {
                result[key] = resolved
            }
        }
        return .object(result)
    }

    private static func applicableOverrides(
        _ raw: JSONValue?,
        for targetContext: ForgeTargetContext
    ) -> [[String: JSONValue]] {
        guard case .object(let object)? = raw else {
            return []
        }

        let keys = targetOverrideKeys(for: targetContext)

        var result: [[String: JSONValue]] = []
        var seen = Set<String>()
        for key in keys where seen.insert(key).inserted {
            guard case .object(let override)? = object[key] else {
                continue
            }
            result.append(override)
        }
        return result
    }

    private static func targetOverrideKeys(for targetContext: ForgeTargetContext) -> [String] {
        let platform = targetContext.platform.trimmingCharacters(in: .whitespacesAndNewlines)
        let formFactor = targetContext.formFactor.trimmingCharacters(in: .whitespacesAndNewlines)
        let surface = targetContext.surface.trimmingCharacters(in: .whitespacesAndNewlines)
        let mobilePlatforms: Set<String> = ["android", "ios"]
        let mobileFormFactors: Set<String> = ["phone", "tablet"]
        let isMobile = mobilePlatforms.contains(platform) || mobileFormFactors.contains(formFactor)
        var keys: [String] = []

        if !surface.isEmpty {
            keys.append("surface:\(surface)")
            keys.append(surface)
        }
        if isMobile {
            keys.append("mobile")
        }
        if !formFactor.isEmpty {
            keys.append("formFactor:\(formFactor)")
            keys.append(formFactor)
        }
        if !platform.isEmpty {
            keys.append(platform)
        }
        if isMobile && !formFactor.isEmpty {
            keys.append("mobile.\(formFactor)")
            keys.append("mobile:\(formFactor)")
            keys.append("mobile/\(formFactor)")
        }
        if !platform.isEmpty && !formFactor.isEmpty {
            keys.append("\(platform).\(formFactor)")
            keys.append("\(platform)/\(formFactor)")
            keys.append("\(platform):\(formFactor)")
        }
        return keys
    }

    private static func matchesTarget(_ raw: JSONValue?, for targetContext: ForgeTargetContext) -> Bool {
        guard let spec = normalizeTarget(raw) else {
            return true
        }

        if !spec.platforms.isEmpty && !spec.platforms.contains(targetContext.platform) {
            return false
        }
        if !spec.excludePlatforms.isEmpty && spec.excludePlatforms.contains(targetContext.platform) {
            return false
        }
        if !spec.formFactors.isEmpty && !spec.formFactors.contains(targetContext.formFactor) {
            return false
        }
        if !spec.capabilities.isEmpty {
            let actual = Set(targetContext.capabilities)
            if !Set(spec.capabilities).isSubset(of: actual) {
                return false
            }
        }
        return true
    }

    private static func normalizeTarget(_ raw: JSONValue?) -> NormalizedTarget? {
        switch raw {
        case nil, .null:
            return nil
        case .string(let value):
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? nil : NormalizedTarget(platforms: [trimmed])
        case .array(let values):
            return normalizedTargetOrNil(NormalizedTarget(platforms: stringList(.array(values))))
        case .object(let object):
            return normalizedTargetOrNil(
                NormalizedTarget(
                    platforms: stringList(object["platforms"]),
                    excludePlatforms: stringList(object["excludePlatforms"]),
                    formFactors: stringList(object["formFactors"]),
                    capabilities: stringList(object["capabilities"])
                )
            )
        default:
            return nil
        }
    }

    private static func normalizedTargetOrNil(_ target: NormalizedTarget) -> NormalizedTarget? {
        guard
            !target.platforms.isEmpty ||
            !target.excludePlatforms.isEmpty ||
            !target.formFactors.isEmpty ||
            !target.capabilities.isEmpty
        else {
            return nil
        }
        return target
    }

    private static func shouldStripTargetingKey(_ key: String, _ value: JSONValue) -> Bool {
        switch key {
        case "target":
            return normalizeTarget(value) != nil
        case "targetOverrides":
            guard case .object(let object) = value else {
                return false
            }
            return !object.isEmpty && object.values.allSatisfy {
                if case .object = $0 {
                    return true
                }
                return false
            }
        default:
            return false
        }
    }

    private static func stringList(_ raw: JSONValue?) -> [String] {
        switch raw {
        case .string(let value):
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? [] : [trimmed]
        case .array(let values):
            return values.compactMap {
                guard case .string(let value) = $0 else { return nil }
                let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
                return trimmed.isEmpty ? nil : trimmed
            }
        default:
            return []
        }
    }

    private static func deepMerge(
        base: [String: JSONValue],
        override: [String: JSONValue]
    ) -> [String: JSONValue] {
        var result = base
        for (key, overrideValue) in override {
            if
                case .object(let currentObject)? = result[key],
                case .object(let overrideObject) = overrideValue
            {
                result[key] = .object(deepMerge(base: currentObject, override: overrideObject))
            } else {
                result[key] = overrideValue
            }
        }
        return result
    }
}

private struct NormalizedTarget {
    let platforms: [String]
    let excludePlatforms: [String]
    let formFactors: [String]
    let capabilities: [String]

    init(
        platforms: [String] = [],
        excludePlatforms: [String] = [],
        formFactors: [String] = [],
        capabilities: [String] = []
    ) {
        self.platforms = platforms
        self.excludePlatforms = excludePlatforms
        self.formFactors = formFactors
        self.capabilities = capabilities
    }
}
